File: 2024/C/day_one.c
#include <stdio.h>
#include <stdlib.h>

#define MAX_SIZE 1000

int compare(const void *a, const void *b) {
    return (*(int *)a - *(int *)b);
}

int load_inputs(const char *filename, int *list_one, int *list_two,
                int max_size, int *rows_loaded) {
    FILE *file;
    int i = 0;

    file = fopen(filename, "r");
    if (file == NULL) {
        perror("Error opening file");
        return -1;
    }

    while (fscanf(file, "%d %d", &list_one[i], &list_two[i]) == 2 &&
           i < max_size) {
        i++;
    }

    fclose(file);
    *rows_loaded = i;
    return i;
}

int calculate_total_distance(int *list_one, int *list_two, int rows) {
    qsort(list_one, rows, sizeof(int), compare);
    qsort(list_two, rows, sizeof(int), compare);

    int total_distance = 0;
    for (int i = 0; i < rows; i++) {
        total_distance += abs(list_one[i] - list_two[i]);
    }

    return total_distance;
}

int calculate_total_similarity(int *list_one, int *list_two, int rows) {
    int total_similarity = 0;

    for (int i = 0; i < rows; i++) {
        int count = 0;
        for (int j = 0; j < rows; j++) {
            if (list_one[i] == list_two[j]) {
                count++;
            }
        }
        total_similarity += list_one[i] * count;
    }

    return total_similarity;
}

int main() {
    int list_one[MAX_SIZE], list_two[MAX_SIZE];
    int rows_loaded = 0;

    load_inputs("../inputs/day_one.txt", list_one, list_two, MAX_SIZE, &rows_loaded);

    if (rows_loaded < 0) {
        fprintf(stderr, "Failed to load data from file.\n");
        return 1;
    }

    int total_distance = calculate_total_distance(list_one, list_two, rows_loaded);
    printf("The Total Distance is %d\n", total_distance);

    int total_similarity = calculate_total_similarity(list_one, list_two, rows_loaded);
    printf("The Toal Similarity is %d\n", total_similarity);

    return 0;
}

File: 2024/Rust/src/day_one.rs
use std::io::Read;

fn load_inputs() -> (Vec<u32>, Vec<u32>) {
    let mut input = String::new();

    let mut file = std::fs::File::open("./inputs/day_one.txt").unwrap();
    file.read_to_string(&mut input).unwrap();

    let mut list_one: Vec<u32> = vec![];
    let mut list_two: Vec<u32> = vec![];

    for line in input.lines() {
        let mut line = line.split_whitespace();
        let first = line.next().unwrap().parse::<u32>().unwrap();
        let second = line.next().unwrap().parse::<u32>().unwrap();
        list_one.push(first);
        list_two.push(second);
    }

    (list_one, list_two)
}

fn calculate_total_distance(mut list_one: Vec<u32>, mut list_two: Vec<u32>) -> u32 {
    list_one.sort();
    list_two.sort();

    list_one
        .iter()
        .zip(list_two.iter())
        .map(|(a, b)| a.abs_diff(*b))
        .sum()
}

fn calculate_total_similarity(list_one: Vec<u32>, list_two: Vec<u32>) -> u32 {
    list_one
        .iter()
        .map(|a| a * list_two.iter().filter(|b| *b == a).count() as u32)
        .sum()
}

pub fn day_one() {
    // Day 1 - Part 1
    let (list_one, list_two) = load_inputs();
    let total = calculate_total_distance(list_one, list_two);
    println!("The Total Distance is {total}");

    // Day 1 - Part 2
    let (list_one, list_two) = load_inputs();
    let similarity = calculate_total_similarity(list_one, list_two);
    println!("The Total Similarity is {similarity}");
}
